### PR TITLE
e2e: test for host volumes and Docker volumes

### PR DESCRIPTION
**Pull request by tgross**
Fixes https://github.com/hashicorp/nomad/issues/8846

Exercises host volume and Docker volume functionality for the `exec` and `docker`
task driver, particularly around mounting locations within the container and
how this can be used with `template`.

---

Note: this test replaces the existing `host_volumes` test, which was pretty much 
a smoke test, because I wanted to show the interaction between components. It
also demonstrates some of the issues in #8962 and #8919

```
$ go test -v . -suite Volumes
=== RUN   TestE2E
...
=== RUN   TestE2E/Volumes
=== RUN   TestE2E/Volumes/*volumes.VolumesTest
=== RUN   TestE2E/Volumes/*volumes.VolumesTest/TestVolumeMounts
--- PASS: TestE2E (13.55s)
...
    --- PASS: TestE2E/Volumes (13.54s)
        --- PASS: TestE2E/Volumes/*volumes.VolumesTest (13.46s)
            --- PASS: TestE2E/Volumes/*volumes.VolumesTest/TestVolumeMounts (13.33s)
PASS
ok      github.com/hashicorp/nomad/e2e  13.669s
```

### CODE REVIEW

File: e2e/volumes/input/volumes.nomad [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+      config {
+        image   = "busybox:1"
+        command = "/bin/sh"
+        args    = ["/usr/local/bin/myapplication.sh"]
+
+        mounts = [
+          # this mount binds the task's own NOMAD_TASK_DIR directory as the
+          # source, letting us map it to a more convenient location; this is a
+          # frequently-used way to get templates into an arbitrary location in
+          # the task for Docker
```
Comment (tgross) [inline e2e/volumes/input/volumes.nomad:24]: h/t to @angrycub for this one; this job should serve as a good starting point to expand the documentation for #8919

File: e2e/e2eutil/allocs.go [hunk 2]
```diff
@@ -144,3 +144,25 @@ func AllocStatusesRescheduled(jobID string) ([]string, error) {
...
+// passed cmd, retrying if the task isn't ready
+func AllocExec(allocID, taskID string, cmd string, wc *WaitConfig) (string, error) {
+
+	//func AllocExec(params string, wc *WaitConfig) (string, error) {
+	var got string
+	var err error
+	interval, retries := wc.OrDefault()
+
+	args := append([]string{"alloc", "exec", "-task", taskID, allocID},
+		strings.Split(cmd, " ")...)
```
Comment (notnoop) [inline e2e/e2eutil/allocs.go:158]: I'd consider using `"/bin/sh", "-c", cmd` here and avoid parsing the command. If the command contains redirects, or quoted strings, or some sh-ism, the result will be somewhat unexpected.

**Review by notnoop — APPROVED** [review]

File: e2e/e2eutil/allocs.go (hunk 2 above)
Comment (tgross) [reply e2e/e2eutil/allocs.go:158]: Good call.

File: e2e/e2eutil/allocs.go [hunk 3]
```diff
@@ -144,3 +144,25 @@ func AllocStatusesRescheduled(jobID string) ([]string, error) {
 	}
 	return statuses, nil
 }
+
+// AllocExec is a convenience wrapper that runs 'nomad alloc exec' with the
+// passed cmd, retrying if the task isn't ready
+func AllocExec(allocID, taskID string, cmd string, wc *WaitConfig) (string, error) {
+
+	//func AllocExec(params string, wc *WaitConfig) (string, error) {
```
[the remark this answers is not in the document]
Comment (tgross) [reply e2e/e2eutil/allocs.go:152]: 🤦 Guess how many times I reworked this function signature?

**Pull request merged**